tidy3d/components/field_projection.py

```diff
@@ -429,96 +425,68 @@ def _far_fields_for_surface(
         theta = np.atleast_1d(theta)
         phi = np.atleast_1d(phi)
 
-        sin_theta = np.sin(theta)
-        cos_theta = np.cos(theta)
-        sin_phi = np.sin(phi)
-        cos_phi = np.cos(phi)
-
-        J = np.zeros((3, len(theta), len(phi)), dtype=complex)
-        M = np.zeros_like(J)
-
-        phase = [None] * 3
         propagation_factor = -1j * AbstractFieldProjectionData.wavenumber(
             medium=medium, frequency=frequency
         )
 
-        def integrate_for_one_theta(i_th: int):
-            """Perform integration for a given theta angle index"""
-
-            for j_ph in np.arange(len(phi)):
-                phase[0] = np.exp(propagation_factor * pts[0] * sin_theta[i_th] * cos_phi[j_ph])
-                phase[1] = np.exp(propagation_factor * pts[1] * sin_theta[i_th] * sin_phi[j_ph])
-                phase[2] = np.exp(propagation_factor * pts[2] * cos_theta[i_th])
-
-                phase_ij = phase[idx_u][:, None] * phase[idx_v][None, :] * phase[idx_w]
-
-                if self.is_2d_simulation:
-                    J[idx_u, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"E{cmp_1}"].values, phase_ij, pts[idx_int_1d]
-                    )
-                    J[idx_v, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"E{cmp_2}"].values, phase_ij, pts[idx_int_1d]
-                    )
-
-                    M[idx_u, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"H{cmp_1}"].values, phase_ij, pts[idx_int_1d]
-                    )
-
-                    M[idx_v, i_th, j_ph] = self.integrate_1d(
-                        currents_f[f"H{cmp_2}"].values, phase_ij, pts[idx_int_1d]
-                    )
-                else:
-                    J[idx_u, i_th, j_ph] = self.integrate_2d(
-                        currents_f[f"E{cmp_1}"].values, phase_ij, pts[idx_u], pts[idx_v]
-                    )
-
-                    J[idx_v, i_th, j_ph] = self.integrate_2d(
-                        currents_f[f"E{cmp_2}"].values, phase_ij, pts[idx_u], pts[idx_v]
-                    )
+        st = np.sin(theta)
+        ct = np.cos(theta)
+        sp = np.sin(phi)
+        cp = np.cos(phi)
+
+        expr = oe.contract_expression(
```

Comment:
I'm finding this part challenging to understand. The original code followed Taflove's textbook, which was more comprehensible. Would it be beneficial to add some explanatory comments here to clarify the logic and purpose of these operations?

Reply:
hey @QimingFlex sorry for the late reply on this. yeah I understand this is quite a big difference to how it was implemented previously. the calculations are the same, so it's still following taflove, but the implementation is obviously quite different.
this is because the calculation is now vectorized over theta and phi. the reason for this is mainly that this is much more efficient during the backward pass, otherwise the memory requirements can go through the roof when looping over each array element.
without gradients, this implementation is around as fast as the original implementation for a small number of theta, and faster for a large number of theta.

because of this vectorization, the einsum makes it much simpler to keep track of the dimensions in those arrays (i.e., it is clearly labeled which dimensions remain, which ones disappear / get summed over). but I understand your concern that this is less readable in terms of just following the computation.

do you think we should make this compromise or should I try to refactor it a bit?

Reply:
Thanks, @yaugenst-flex. I've reviewed the operations, and they look good to me! It's great to see some acceleration achieved for the frontend Near2far. I do have one suggestion: it might be better to keep with the original abbreviations, such as `phase` instead of `ehp`, as the former is more intuitive. The same applies to terms like `st` and `ct`. Although the original names are longer, they are generally easier to follow.

Reply:
thanks! yeah that's a good points, I implemented your suggestions as well as some other readability improvements as well